### PR TITLE
Nifi integration revision

### LINKED ISSUE #593: Apache NiFi integration should allow Expression Language
Apache NiFi integration should allow us to use Expression Language for PLC connection string and PLC resource address string.

We sometimes need to get data from 100 different addresses. Current processors doesn't allow us to create those strings on the fly. So we need to enter them manually or we need utilize NiFi API to create flows automatically. 

If those parameters allows us to use expression language, we can read a list from csv file or from database and then we can read them in a loop in Apache NiFi.  

So it will be very handy feature to dynamically specify the connection string and the address string.

For starting point PutFile processor can be examined. This processor utilizes expression language for it's "Directory" parameter.

[https://github.com/apache/nifi/blob/master/nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutFile.java](https://github.com/apache/nifi/blob/master/nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutFile.java)

Notable Lines:

import org.apache.nifi.expression.ExpressionLanguageScope;

...

Define the parameter as expression language supported.

public static final PropertyDescriptor DIRECTORY = new PropertyDescriptor.Builder()
 .name("Directory")
 .description("The directory to which files should be written. You may use expression language such as /aa/bb/${path}")
 .required(true)
 .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
 *.expressionLanguageSupported(ExpressionLanguageScope.FLOWFILE_ATTRIBUTES)*
 .build();

...

And then script can be evaluated in the onTrigger event like

context.getProperty(DIRECTORY).evaluateAttributeExpressions(flowFile).getValue()

Imported from Jira [PLC4X-196](https://issues.apache.org/jira/browse/PLC4X-196). Original Jira may contain additional context.
Reported by: turker.tunali.

### LINKED ISSUE #629: Nifi  BasePlc4xProcessor define validation logic for PLC_CONNECTION_STRING and PLC_ADDRESS_STRING properties
Apache Nifi integration, define logic for the validation methods included in the class BasePlc4xProcessor, for the properties PLC_CONNECTION_STRING and PLC_ADDRESS_STRING

Imported from Jira [PLC4X-328](https://issues.apache.org/jira/browse/PLC4X-328). Original Jira may contain additional context.
Reported by: inigoao.

### PR DESCRIPTION
**Pull request by QuanticPony**
# NiFi integration revision
As we aim for a new release I bring a revision of the NiFi integration with a couple of improvements:

* #900: this made the processors be able to auto-reconnect with small changes. And allowed to make:
    * #593 : this was already done for the addresses. But the connection string had to be constant. Expression Language is now supported on the connection string.
    * #629 : this was an old TODO. I have added validation for the connection string (if not EL in use) and to the addresses (if not EL in use in neither of them as we need the driver).

* Add FilePropertyAccessStrategy: now we can read the addresses from a file in JSON format. Same as 'Address Text' property but in a file. Expression Language can be used and validation for the tags also works.
*  Processors code revision: rewrote the onTrigger methods of the processors so they look more similar between them. To make them more maintainable.
* Complete FlowFlile transfer: some FlowFiles did not transfer to failure correctly. Now the incoming FlowFile is sent to failure and one attribute is added with the localized message in case of any exception.
* Timestamp field name: It was a fixed `ts` field/attribute we added. I have made a property that defines this field's name. 
* Update Readme to reflect changes.

Closes #593
Closes #629

### CODE REVIEW

**Comment by chrisdutz** [comment]
Thanks for this ;-) 

Unfortunately I am probably not the one ideal to review NiFi related things, but I would trust you and your changes. So if I don't hear any objections till Monday, I'll merge this PR into develop before cutting the release branch.

**Comment by chrisdutz** [comment]
By the way ... would be cool, if you could sign up to the dev-list and start communicating with us there ;-)

**Review by ottobackwards — CHANGES_REQUESTED** [review]
Thanks for the work!

- There is a lot of duplication of methods between each of the processors.  These should be moved so that we don't have multiple copies in each processor
- I would like to see more tests for the new evaluation enabled things
- I would like to see tests around the caching with evaluations
- having properties read from files is not a great pattern for nifi, and there are no tests for it

in short, please add tests for all the new things and things that are now effected by the permutations possible.  Please de-duplicate the methods in the various processors.

In the future, smaller pr's would make it easier to review

**Comment by QuanticPony** [comment]
Sorry for the long PR. I would normally split it, but though posting as early as possible would be best in this case.
Duplicated code removed. Unit tests added.
i added the read tags from a file as I have seen it in some production environments with other OPCUA processors. I think it could be easier to migrate to PLC4X if we have them. If you don't see it clear i can remove it

**Comment by QuanticPony** [comment]
I'm done. Tell me if anything else is needed. Will have a look at it over the weekend

**Comment by TurkerTunali** [comment]
So, should we try the NiFi integration now?

**Comment by chrisdutz** [comment]
Well it should be part of the Release ... so feel free to try.
